### ISSUE TITLE
CUDA multi-GPU --device bug

### ISSUE THREAD

**Issue by huangfeng95** [ultralytics/yolov5#1695]
## ❔Question
I specified parameters device but could not work.
Can you please tell me what to troubleshoot?

## Additional context
![image](https://user-images.githubusercontent.com/19809051/102223473-b77eab00-3f1f-11eb-9c0e-571275217bed.png)

**Comment by glenn-jocher** [comment]
@huangfeng95 yes there is a bug in our --device selection method now in more recent versions of torch it seems. Current  workaround is to can define visible devices before running a command:

```bash
$ CUDA_VISIBLE_DEVICES=2,3 python train.py ...

or 

$ export CUDA_VISIBLE_DEVICES=2,3
$ python train.py ...
```

TODO: CUDA multi-gpu --device bug.

**Comment by glenn-jocher** [comment]
@NanoCode012 do you have any insight into this bug? The only thing I can think of is that the torch functionality changed in a recent release, and our current method of setting CUDA_VISIBLE_DEVICES right before checking devices no longer works. The relevant function is here, I'm kind of empty on ideas for a fix.

https://github.com/ultralytics/yolov5/blob/e92245a38c90be884a05cb1e3fe325037b9f1379/utils/torch_utils.py#L46-L69

**Comment by glenn-jocher** [comment]
```python
>>> import os
>>> os.environ['CUDA_VISIBLE_DEVICES']
'0,1,2'
>>> os.environ['CUDA_VISIBLE_DEVICES'] = '0,1'
>>> os.environ['CUDA_VISIBLE_DEVICES']
'0,1'
>>> torch.cuda.device_count()
3
>>> del(torch)
>>> import torch
>>> torch.cuda.device_count()
3
>>> torch.cuda.init()
>>> torch.cuda.device_count()
3
```

**Comment by NanoCode012** [comment]
@glenn-jocher I'm testing my env right now. Both shows the same problem. I think something might have happened to code base then because I have not updated any of my environments for quite a while.

- py 3.7 and torch 1.6
- docker py 3.6 and torch 1.7

Edit: Hm, running the above from code terminal produced confusing results. Will go test some more.

**Comment by NanoCode012** [comment]
@glenn-jocher , good news (?) ! I tracked down the commit which caused this error (by manually going through commits daily.) Luckily, it was only a few days ago. It was the Profile PR ada90e3901da9d24f88e4d1378be96265770a932 .

To reproduce:
```bash
# Working on 11th Dec
git checkout -b 11a 94a7f55c4e5cca3dfe4de0bd0793173d5b152ec5
python train.py --img 640 --batch 16 --epochs 3 --data coco128.yaml --weights yolov5s.pt --nosave --cache --device 0,1

# Error on 12th Dec
git checkout -b 12 ada90e3901da9d24f88e4d1378be96265770a932
python train.py --img 640 --batch 16 --epochs 3 --data coco128.yaml --weights yolov5s.pt --nosave --cache --device 0,1
```

I am not sure what exactly in this PR that caused this change because it seems so isolated.

After some tests, it was this. You assigned the device in the function declaration.. I'll submit a PR to this..
https://github.com/ultralytics/yolov5/blob/69ea70cd3b4c3e908bbfac094bc4550a3dcf43e3/utils/torch_utils.py#L78

**Comment by NanoCode012** [comment]
@huangfeng95 @glenn-jocher , could you please test the PR and tell whether it worked for you? 

I tested that this change worked on both of my environment above. What a nasty bug!

```bash
git clone https://github.com/NanoCode012/yolov5.git -b gpu-fix && cd yolov5
python train.py --epochs 3 --device 0 # Single
python train.py --epochs 3 --device 0,1 # DP
python -m torch.distributed.launch --nproc_per_node 2 train.py --device 0,1 # DDP
```

**Comment by glenn-jocher** [comment]
@NanoCode012 wow!! We should give you a medal for finding this bug! 

That's very bizarre. The profile() function is not actually used anywhere during normal repo operation, it's only there for manual use when comparing two modules, for example when developing new architectures. I used it here to evaluate a Focus() module alternative https://github.com/ultralytics/yolov5/issues/1274#issuecomment-743341921. It does all the dirty work for you and shows you the time impact forward backward.

I'll test the PR.

**Comment by glenn-jocher** [comment]
@huangfeng95 this bug should be resolved now thanks to PR #1701 by @NanoCode012. Please git pull to receive this update and let us know if you have any other problems!
